compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -522,11 +522,11 @@ pub enum TerminatorKind<'tcx> {
     /// deaggregation runs.
     Resume,
 
-    /// Indicates that the landing pad is finished and that the process should abort.
+    /// Indicates that the landing pad is finished and that the process should terminate.
     ///
     /// Used to prevent unwinding for foreign items or with `-C unwind=abort`. Only permitted in
     /// cleanup blocks.
-    Abort,
+    Terminate,
```

Comment:
So now we have a bunch of terminators and one of them is called `Terminate`? That seems odd. `Abort`  sounds more descriptive? Why the rename?

Reply:
There was a suggestion to keep the ` UnwindAction` variant name and terminator name the same, so I renamed the terminator name `Terminate`. `UnwindAction::Terminate` was named `terminate` because it's the well-established name for unwinding handling. IMO, `Abort` wasn't a good name anyway because it might be confused with `std::process::abort` or `core::intrinsics::abort`.

I am open to suggestions of better names.

Reply:
>  IMO, Abort wasn't a good name anyway because it might be confused with std::process::abort or core::intrinsics::abort.

It is a lot like both of these, so -- what is the problem here?

Reply:
It's codegened to a call of `panic_no_unwind`.

Reply:
Then the Miri implementation needs to be updated. It used to be codegen'd to an abort intrinsic, I think?

Also sounds like the name should really be more specific then. `PanicNoUnwind` or so.

Reply:
I don't think we want to specify how exactly this terminator and `UnwindAction::Terminate` behaves. Its behaviour can change as long as it's terminating execution. For example, `UnwindAction::Terminate` cause an immediate abort on MSVC SEH when it happens in cleanup. There're also platforms where we delegate exception handling to C++ and `std::terminate()` might be called in such case. I don't see it as an issue that for MIRI it triggers an abort with `panic in a function that cannot unwind` abort message (this is the message used by `panic_no_unwind` by the way)

Reply:
Calling `panic_no_unwind` means we should only use this terminator when this is really the situation -- a panic occurred in a function that is not allowed to unwind. There is a certain intent associated with this. It would not make sense, for example, to implement `core::intrinsics::abort()` with this terminator.

I would expect an `Abort` or `Terminate` terminator to be much more broadly usable. Compiling  `core::intrinsics::abort()`  to a terminator of that name sounds totally reasonable.

I think there are two coherent option here:
- Either this is a general-purpose "stop the execution" terminator. Then both the error message and the name of the terminator should reflect that.
- Or it is specific to the situation of panicing in a function that must not unwind. Then both the error message and the name of the terminator should reflect that.

Currently the name indicates the former but the message indicates the latter, and that's not good.

Reply:
>  I don't see it as an issue that for MIRI it triggers an abort with panic in a function that cannot unwind abort message (this is the message used by panic_no_unwind by the way)

That is not what it does in Miri though. In Miri the terminator aborts saying 'the program aborted execution'. (See `compiler/rustc_const_eval/src/interpret/terminator.rs`.)

Reply:
It was always the latter. Even when it's implemented `core::intrinsics::abort()` the comment says `Indicates that the landing pad is finished and that the process should abort.` and we have checks in place to ensure that this is only ever used in cleanup blocks.

My point is that `PanicNoUnwind` is not a good name because we only guarantee program termination, without guaranteeing that the panic handler will be invoked. The name should hint that it's illegal for us to just generate it as `abort`.

Reply:
> That is not what `Terminate` does in Miri though. In Miri the terminator aborts saying 'the program aborted execution'. (See `compiler/rustc_const_eval/src/interpret/terminator.rs`.) In codegen it does something else. This makes no sense.

I neglected that one. Fixed.

Reply:
> My point is that PanicNoUnwind is not a good name because we only guarantee program termination, without guaranteeing that the panic handler will be invoked.

Oh I see. Maybe something like `TerminateCannotUnwind`? `AfterCleanupTerminate`? `TerminateAfterCleanup`? Or maybe `Resume` and `Abort`/`Terminate` should be merged into `CleanupFinished` with a parameter indicating whether to resume or terminate?

Reply:
I think we can leave name bikeshedding / merging `Resume`/`Terminate` to a follow up PR.

Reply:
I assume it'll never happen then... this PR doesn't make it worse, but it touches the name anyway. :shrug: